File: packages/python/scalix/providers/local/direct_llm.py
"""Direct LLM provider — calls OpenAI/Anthropic/Google APIs directly.

Used in local mode when no Scalix API key is configured.
The developer uses their own provider API keys.
"""

from __future__ import annotations

import json
import logging
from typing import Any, AsyncIterator

from scalix.config import ScalixConfig, get_config
from scalix.exceptions import ConfigurationError, ModelNotFoundError
from scalix.providers.base import LLMProvider
from scalix.types import Message, Role, StreamEvent, StreamEventType, ToolCall

logger = logging.getLogger("scalix.providers.direct_llm")


class DirectLLM(LLMProvider):
    """Direct LLM API calls using the developer's own API keys.

    Supports OpenAI, Anthropic, Google, and Ollama.
    No Scalix infrastructure involved — zero cost to Scalix.
    """

    def __init__(self, config: ScalixConfig | None = None) -> None:
        self._config = config or get_config()
        self._openai_client: Any = None
        self._anthropic_client: Any = None

    async def chat(
        self,
        messages: list[Message],
        model: str,
        tools: list[dict[str, Any]] | None = None,
        temperature: float = 0.7,
        stream: bool = False,
        **kwargs: Any,
    ) -> Message:
        """Route to the appropriate provider based on model name."""
        provider = self._detect_provider(model)

        if provider == "openai":
            return await self._call_openai(messages, model, tools, temperature, **kwargs)
        elif provider == "anthropic":
            return await self._call_anthropic(messages, model, tools, temperature, **kwargs)
        elif provider == "google":
            return await self._call_google(messages, model, tools, temperature, **kwargs)
        elif provider == "ollama":
            return await self._call_ollama(messages, model, tools, temperature, **kwargs)
        else:
            raise ConfigurationError(
                f"Cannot determine provider for model '{model}'. "
                "Set OPENAI_API_KEY, ANTHROPIC_API_KEY, or GOOGLE_API_KEY."
            )

    async def stream_chat(
        self,
        messages: list[Message],
        model: str,
        tools: list[dict[str, Any]] | None = None,
        temperature: float = 0.7,
        **kwargs: Any,
    ) -> AsyncIterator[StreamEvent]:
        """Stream responses from the appropriate provider."""
        provider = self._detect_provider(model)

        if provider == "openai":
            async for event in self._stream_openai(messages, model, tools, temperature, **kwargs):
                yield event
        elif provider == "anthropic":
            async for event in self._stream_anthropic(
                messages, model, tools, temperature, **kwargs
            ):
                yield event
        else:
            raise ConfigurationError(f"Streaming not yet supported for provider '{provider}'")

    def _detect_provider(self, model: str) -> str:
        """Detect which provider a model belongs to."""
        model_lower = model.lower()

        if any(m in model_lower for m in ["gpt", "o1", "o3", "o4"]):
            return "openai"
        elif any(m in model_lower for m in ["claude", "haiku", "sonnet", "opus"]):
            return "anthropic"
        elif any(m in model_lower for m in ["gemini", "palm"]):
            return "google"
        elif self._config.ollama_host:
            return "ollama"
        elif model == "auto":
            # Auto-select based on available keys
            if self._config.anthropic_api_key:
                return "anthropic"
            elif self._config.openai_api_key:
                return "openai"
            elif self._config.google_api_key:
                return "google"
        elif self._config.openai_api_key:
            return "openai"
        elif self._config.anthropic_api_key:
            return "anthropic"
        elif self._config.google_api_key:
            return "google"

        raise ConfigurationError(
            "No LLM provider configured. Set one of: "
            "OPENAI_API_KEY, ANTHROPIC_API_KEY, GOOGLE_API_KEY, or OLLAMA_HOST"
        )

    def _resolve_model(self, model: str, provider: str) -> str:
        """Resolve 'auto' model to a concrete model name."""
        if model != "auto":
            return model

        defaults = {
            "openai": "gpt-4o",
            "anthropic": "claude-sonnet-4-20250514",
            "google": "gemini-2.5-flash",
            "ollama": "llama3.2",
        }
        return defaults.get(provider, model)

    # --- OpenAI ---

    def _get_openai_client(self) -> Any:
        if self._openai_client is None:
            try:
                from openai import AsyncOpenAI
            except ImportError:
                raise ConfigurationError(
                    "OpenAI package not installed. Run: pip install scalix[openai]"
                )
            kwargs: dict[str, Any] = {"api_key": self._config.openai_api_key}
            if self._config.openai_base_url:
                kwargs["base_url"] = self._config.openai_base_url
            self._openai_client = AsyncOpenAI(**kwargs)
        return self._openai_client

    async def _call_openai(
        self,
        messages: list[Message],
        model: str,
        tools: list[dict[str, Any]] | None,
        temperature: float,
        **kwargs: Any,
    ) -> Message:
        client = self._get_openai_client()
        model = self._resolve_model(model, "openai")

        openai_messages = self._to_openai_messages(messages)
        openai_tools = self._to_openai_tools(tools) if tools else None

        request: dict[str, Any] = {
            "model": model,
            "messages": openai_messages,
            "temperature": temperature,
        }
        if openai_tools:
            request["tools"] = openai_tools

        logger.debug("OpenAI request: model=%s, messages=%d", model, len(openai_messages))
        response = await client.chat.completions.create(**request)
        choice = response.choices[0]

        # Extract tool calls if present
        tool_calls: list[ToolCall] | None = None
        if choice.message.tool_calls:
            tool_calls = [
                ToolCall(
                    id=tc.id,
                    name=tc.function.name,
                    arguments=json.loads(tc.function.arguments),
                )
                for tc in choice.message.tool_calls
            ]

        return Message(
            role=Role.ASSISTANT,
            content=choice.message.content or "",
            tool_calls=tool_calls,
        )

    async def _stream_openai(
        self,
        messages: list[Message],
        model: str,
        tools: list[dict[str, Any]] | None,
        temperature: float,
        **kwargs: Any,
    ) -> AsyncIterator[StreamEvent]:
        client = self._get_openai_client()
        model = self._resolve_model(model, "openai")

        request: dict[str, Any] = {
            "model": model,
            "messages": self._to_openai_messages(messages),
            "temperature": temperature,
            "stream": True,
        }
        if tools:
            request["tools"] = self._to_openai_tools(tools)

        stream = await client.chat.completions.create(**request)
        async for chunk in stream:
            delta = chunk.choices[0].delta if chunk.choices else None
            if delta and delta.content:
                yield StreamEvent(type=StreamEventType.TEXT_DELTA, data=delta.content)

        yield StreamEvent(type=StreamEventType.DONE)

    def _to_openai_messages(self, messages: list[Message]) -> list[dict[str, Any]]:
        result = []
        for msg in messages:
            entry: dict[str, Any] = {"role": msg.role.value, "content": msg.content}
            if msg.tool_call_id:
                entry["tool_call_id"] = msg.tool_call_id
            if msg.tool_calls:
                entry["tool_calls"] = [
                    {
                        "id": tc.id,
                        "type": "function",
                        "function": {
                            "name": tc.name,
                            "arguments": json.dumps(tc.arguments),
                        },
                    }
                    for tc in msg.tool_calls
                ]
            if msg.name:
                entry["name"] = msg.name
            result.append(entry)
        return result

    def _to_openai_tools(self, tools: list[dict[str, Any]]) -> list[dict[str, Any]]:
        return [
            {
                "type": "function",
                "function": {
                    "name": tool["name"],
                    "description": tool.get("description", ""),
                    "parameters": tool.get("parameters", {}),
                },
            }
            for tool in tools
        ]

    # --- Anthropic ---

    def _get_anthropic_client(self) -> Any:
        if self._anthropic_client is None:
            try:
                from anthropic import AsyncAnthropic
            except ImportError:
                raise ConfigurationError(
                    "Anthropic package not installed. Run: pip install scalix[anthropic]"
                )
            kwargs: dict[str, Any] = {"api_key": self._config.anthropic_api_key}
            if self._config.anthropic_base_url:
                kwargs["base_url"] = self._config.anthropic_base_url
            self._anthropic_client = AsyncAnthropic(**kwargs)
        return self._anthropic_client

    async def _call_anthropic(
        self,
        messages: list[Message],
        model: str,
        tools: list[dict[str, Any]] | None,
        temperature: float,
        **kwargs: Any,
    ) -> Message:
        client = self._get_anthropic_client()
        model = self._resolve_model(model, "anthropic")

        system_prompt, anthropic_messages = self._to_anthropic_messages(messages)
        anthropic_tools = self._to_anthropic_tools(tools) if tools else None

        request: dict[str, Any] = {
            "model": model,
            "messages": anthropic_messages,
            "max_tokens": kwargs.get("max_tokens", 4096),
            "temperature": temperature,
        }
        if system_prompt:
            request["system"] = system_prompt
        if anthropic_tools:
            request["tools"] = anthropic_tools

        logger.debug("Anthropic request: model=%s, messages=%d", model, len(anthropic_messages))
        response = await client.messages.create(**request)

        # Parse response — may contain text and/or tool_use blocks
        content_text = ""
        tool_calls: list[ToolCall] = []

        for block in response.content:
            if block.type == "text":
                content_text += block.text
            elif block.type == "tool_use":
                tool_calls.append(
                    ToolCall(
                        id=block.id,
                        name=block.name,
                        arguments=block.input if isinstance(block.input, dict) else {},
                    )
                )

        return Message(
            role=Role.ASSISTANT,
            content=content_text,
            tool_calls=tool_calls if tool_calls else None,
        )

    async def _stream_anthropic(
        self,
        messages: list[Message],
        model: str,
        tools: list[dict[str, Any]] | None,
        temperature: float,
        **kwargs: Any,
    ) -> AsyncIterator[StreamEvent]:
        client = self._get_anthropic_client()
        model = self._resolve_model(model, "anthropic")

        system_prompt, anthropic_messages = self._to_anthropic_messages(messages)

        request: dict[str, Any] = {
            "model": model,
            "messages": anthropic_messages,
            "max_tokens": kwargs.get("max_tokens", 4096),
            "temperature": temperature,
        }
        if system_prompt:
            request["system"] = system_prompt
        if tools:
            request["tools"] = self._to_anthropic_tools(tools)

        async with client.messages.stream(**request) as stream:
            async for text in stream.text_stream:
                yield StreamEvent(type=StreamEventType.TEXT_DELTA, data=text)

        yield StreamEvent(type=StreamEventType.DONE)

    def _to_anthropic_messages(
        self, messages: list[Message]
    ) -> tuple[str, list[dict[str, Any]]]:
        """Convert to Anthropic format. Extract system prompt separately."""
        system_prompt = ""
        anthropic_messages: list[dict[str, Any]] = []

        for msg in messages:
            if msg.role == Role.SYSTEM:
                system_prompt = msg.content
                continue

            if msg.role == Role.TOOL:
                anthropic_messages.append(
                    {
                        "role": "user",
                        "content": [
                            {
                                "type": "tool_result",
                                "tool_use_id": msg.tool_call_id,
                                "content": msg.content,
                            }
                        ],
                    }
                )
            elif msg.role == Role.ASSISTANT and msg.tool_calls:
                content: list[dict[str, Any]] = []
                if msg.content:
                    content.append({"type": "text", "text": msg.content})
                for tc in msg.tool_calls:
                    content.append(
                        {
                            "type": "tool_use",
                            "id": tc.id,
                            "name": tc.name,
                            "input": tc.arguments,
                        }
                    )
                anthropic_messages.append({"role": "assistant", "content": content})
            else:
                anthropic_messages.append(
                    {"role": msg.role.value, "content": msg.content}
                )

        return system_prompt, anthropic_messages

    def _to_anthropic_tools(self, tools: list[dict[str, Any]]) -> list[dict[str, Any]]:
        return [
            {
                "name": tool["name"],
                "description": tool.get("description", ""),
                "input_schema": tool.get("parameters", {"type": "object", "properties": {}}),
            }
            for tool in tools
        ]

    # --- Google ---

    async def _call_google(
        self,
        messages: list[Message],
        model: str,
        tools: list[dict[str, Any]] | None,
        temperature: float,
        **kwargs: Any,
    ) -> Message:
        """Call Google Generative AI API."""
        try:
            import google.generativeai as genai
        except ImportError:
            raise ConfigurationError(
                "Google AI package not installed. Run: pip install scalix[google]"
            )

        genai.configure(api_key=self._config.google_api_key)
        model_name = self._resolve_model(model, "google")
        gen_model = genai.GenerativeModel(model_name)

        # Build simple prompt from messages
        prompt_parts = []
        for msg in messages:
            prefix = {"system": "System: ", "user": "", "assistant": "Assistant: ", "tool": "Tool result: "}
            prompt_parts.append(f"{prefix.get(msg.role.value, '')}{msg.content}")

        response = await gen_model.generate_content_async(
            "\n\n".join(prompt_parts),
            generation_config={"temperature": temperature},
        )

        return Message(role=Role.ASSISTANT, content=response.text or "")

    # --- Ollama ---

    async def _call_ollama(
        self,
        messages: list[Message],
        model: str,
        tools: list[dict[str, Any]] | None,
        temperature: float,
        **kwargs: Any,
    ) -> Message:
        """Call local Ollama instance via OpenAI-compatible API."""
        try:
            from openai import AsyncOpenAI
        except ImportError:
            raise ConfigurationError(
                "OpenAI package needed for Ollama compatibility. Run: pip install scalix[openai]"
            )

        host = self._config.ollama_host or "http://localhost:11434"
        client = AsyncOpenAI(base_url=f"{host}/v1", api_key="ollama")
        model = self._resolve_model(model, "ollama")

        response = await client.chat.completions.create(
            model=model,
            messages=self._to_openai_messages(messages),
            temperature=temperature,
        )

        choice = response.choices[0]
        return Message(role=Role.ASSISTANT, content=choice.message.content or "")

File: packages/typescript/src/providers/local/directLlm.ts
/**
 * Direct LLM provider — calls OpenAI/Anthropic APIs directly.
 *
 * Used in local mode when no Scalix API key is configured.
 * The developer uses their own provider API keys.
 */

import { getConfig, dynamicImport, type ScalixConfig } from '../../config.js';
import { ConfigurationError } from '../../errors.js';
import type { LLMProvider } from '../base.js';
import type { Message, StreamEvent, ToolCall } from '../../types.js';

type ProviderName = 'openai' | 'anthropic' | 'google' | 'ollama';

/**
 * Direct LLM API calls using the developer's own API keys.
 *
 * Supports OpenAI, Anthropic, Google, and Ollama.
 * No Scalix infrastructure involved — zero cost to Scalix.
 */
export class DirectLLM implements LLMProvider {
  private config: ScalixConfig;
  private openaiClient: unknown = null;
  private anthropicClient: unknown = null;

  constructor(config?: ScalixConfig) {
    this.config = config ?? getConfig();
  }

  async chat(
    messages: Message[],
    model: string,
    options?: {
      tools?: Record<string, unknown>[];
      temperature?: number;
      stream?: boolean;
    },
  ): Promise<Message> {
    const provider = this.detectProvider(model);
    const temperature = options?.temperature ?? 0.7;
    const tools = options?.tools;

    switch (provider) {
      case 'openai':
        return this.callOpenAI(messages, model, tools, temperature);
      case 'anthropic':
        return this.callAnthropic(messages, model, tools, temperature);
      case 'google':
        return this.callGoogle(messages, model, temperature);
      case 'ollama':
        return this.callOllama(messages, model, temperature);
      default:
        throw new ConfigurationError(
          `Cannot determine provider for model '${model}'. ` +
            'Set OPENAI_API_KEY, ANTHROPIC_API_KEY, or GOOGLE_API_KEY.',
        );
    }
  }

  async *streamChat(
    messages: Message[],
    model: string,
    options?: {
      tools?: Record<string, unknown>[];
      temperature?: number;
    },
  ): AsyncGenerator<StreamEvent> {
    const provider = this.detectProvider(model);
    const temperature = options?.temperature ?? 0.7;

    if (provider === 'openai') {
      yield* this.streamOpenAI(messages, model, temperature);
    } else if (provider === 'anthropic') {
      yield* this.streamAnthropic(messages, model, temperature);
    } else {
      throw new ConfigurationError(`Streaming not yet supported for provider '${provider}'`);
    }
  }

  // --- Provider Detection ---

  private detectProvider(model: string): ProviderName {
    const m = model.toLowerCase();

    // Scalix-branded models require a Scalix API key (cloud mode).
    // In local mode, fall through to BYOK provider detection.
    if (m.startsWith('scalix-') || m === 'auto') {
      if (this.config.googleApiKey) return 'google';
      if (this.config.anthropicApiKey) return 'anthropic';
      if (this.config.openaiApiKey) return 'openai';
      if (this.config.ollamaHost) return 'ollama';

      throw new ConfigurationError(
        'Scalix models require a Scalix API key (cloud mode) or a local provider key. ' +
          'Call configure({ apiKey: "sk_scalix_..." }) or set SCALIX_API_KEY.',
      );
    }

    // BYOK: detect provider from raw model name
    if (['gpt', 'o1', 'o3', 'o4'].some((p) => m.includes(p))) return 'openai';
    if (['claude', 'haiku', 'sonnet', 'opus'].some((p) => m.includes(p))) return 'anthropic';
    if (['gemini', 'palm'].some((p) => m.includes(p))) return 'google';
    if (this.config.ollamaHost) return 'ollama';

    if (this.config.openaiApiKey) return 'openai';
    if (this.config.anthropicApiKey) return 'anthropic';
    if (this.config.googleApiKey) return 'google';

    throw new ConfigurationError(
      'No LLM provider configured. Set SCALIX_API_KEY for cloud mode, ' +
        'or set OPENAI_API_KEY, ANTHROPIC_API_KEY, GOOGLE_API_KEY, or OLLAMA_HOST for local mode.',
    );
  }

  private resolveModel(model: string, provider: ProviderName): string {
    // Map Scalix-branded names to actual provider models for local mode
    const scalixLocalDefaults: Record<string, Record<ProviderName, string>> = {
      'scalix-world-ai': {
        openai: 'gpt-4o-mini',
        anthropic: 'claude-haiku-4-5-20251001',
        google: 'gemini-2.5-flash-lite',
        ollama: 'llama3.2',
      },
      'scalix-advanced': {
        openai: 'gpt-4o',
        anthropic: 'claude-sonnet-4-20250514',
        google: 'gemini-2.5-pro',
        ollama: 'llama3.2',
      },
      'auto': {
        openai: 'gpt-4o-mini',
        anthropic: 'claude-haiku-4-5-20251001',
        google: 'gemini-2.5-flash-lite',
        ollama: 'llama3.2',
      },
    };

    const mapping = scalixLocalDefaults[model.toLowerCase()];
    if (mapping) return mapping[provider] ?? model;

    return model;
  }

  // --- OpenAI ---

  private async getOpenAIClient(): Promise<unknown> {
    if (!this.openaiClient) {
      try {
        const { OpenAI } = await dynamicImport('openai') as { OpenAI: new (opts: { apiKey?: string; baseURL?: string }) => unknown };
        this.openaiClient = new OpenAI({
          apiKey: this.config.openaiApiKey,
          ...(this.config.openaiBaseUrl && { baseURL: this.config.openaiBaseUrl }),
        });
      } catch {
        throw new ConfigurationError(
          'OpenAI package not installed. Run: npm install openai',
        );
      }
    }
    return this.openaiClient;
  }

  private async callOpenAI(
    messages: Message[],
    model: string,
    tools: Record<string, unknown>[] | undefined,
    temperature: number,
  ): Promise<Message> {
    const client = await this.getOpenAIClient() as {
      chat: {
        completions: {
          create: (req: Record<string, unknown>) => Promise<{
            choices: Array<{
              message: {
                content: string | null;
                tool_calls?: Array<{
                  id: string;
                  function: { name: string; arguments: string };
                }>;
              };
            }>;
          }>;
        };
      };
    };

    const resolvedModel = this.resolveModel(model, 'openai');
    const openaiMessages = this.toOpenAIMessages(messages);

    const request: Record<string, unknown> = {
      model: resolvedModel,
      messages: openaiMessages,
      temperature,
    };

    if (tools?.length) {
      request.tools = this.toOpenAITools(tools);
    }

    const response = await client.chat.completions.create(request);
    const choice = response.choices[0];

    let toolCalls: ToolCall[] | undefined;
    if (choice.message.tool_calls?.length) {
      toolCalls = choice.message.tool_calls.map((tc) => ({
        id: tc.id,
        name: tc.function.name,
        arguments: JSON.parse(tc.function.arguments),
      }));
    }

    return {
      role: 'assistant',
      content: choice.message.content ?? '',
      toolCalls,
    };
  }

  private async *streamOpenAI(
    messages: Message[],
    model: string,
    temperature: number,
  ): AsyncGenerator<StreamEvent> {
    const client = await this.getOpenAIClient() as {
      chat: {
        completions: {
          create: (req: Record<string, unknown>) => Promise<AsyncIterable<{
            choices: Array<{ delta?: { content?: string } }>;
          }>>;
        };
      };
    };

    const resolvedModel = this.resolveModel(model, 'openai');

    const stream = await client.chat.completions.create({
      model: resolvedModel,
      messages: this.toOpenAIMessages(messages),
      temperature,
      stream: true,
    });

    for await (const chunk of stream as AsyncIterable<{
      choices: Array<{ delta?: { content?: string } }>;
    }>) {
      const delta = chunk.choices[0]?.delta;
      if (delta?.content) {
        yield { type: 'text_delta', data: delta.content };
      }
    }

    yield { type: 'done' };
  }

  private toOpenAIMessages(messages: Message[]): Record<string, unknown>[] {
    return messages.map((msg) => {
      const entry: Record<string, unknown> = { role: msg.role, content: msg.content };
      if (msg.toolCallId) entry.tool_call_id = msg.toolCallId;
      if (msg.name) entry.name = msg.name;
      if (msg.toolCalls) {
        entry.tool_calls = msg.toolCalls.map((tc) => ({
          id: tc.id,
          type: 'function',
          function: { name: tc.name, arguments: JSON.stringify(tc.arguments) },
        }));
      }
      return entry;
    });
  }

  private toOpenAITools(tools: Record<string, unknown>[]): Record<string, unknown>[] {
    return tools.map((tool) => ({
      type: 'function',
      function: {
        name: tool.name,
        description: tool.description ?? '',
        parameters: tool.parameters ?? {},
      },
    }));
  }

  // --- Anthropic ---

  private async getAnthropicClient(): Promise<unknown> {
    if (!this.anthropicClient) {
      try {
        const { default: Anthropic } = await dynamicImport('@anthropic-ai/sdk') as { default: new (opts: { apiKey?: string; baseURL?: string }) => unknown };
        this.anthropicClient = new Anthropic({
          apiKey: this.config.anthropicApiKey,
          ...(this.config.anthropicBaseUrl && { baseURL: this.config.anthropicBaseUrl }),
        });
      } catch {
        throw new ConfigurationError(
          'Anthropic package not installed. Run: npm install @anthropic-ai/sdk',
        );
      }
    }
    return this.anthropicClient;
  }

  private async callAnthropic(
    messages: Message[],
    model: string,
    tools: Record<string, unknown>[] | undefined,
    temperature: number,
  ): Promise<Message> {
    const client = await this.getAnthropicClient() as {
      messages: {
        create: (req: Record<string, unknown>) => Promise<{
          content: Array<
            | { type: 'text'; text: string }
            | { type: 'tool_use'; id: string; name: string; input: Record<string, unknown> }
          >;
        }>;
      };
    };

    const resolvedModel = this.resolveModel(model, 'anthropic');
    const { systemPrompt, messages: anthropicMessages } = this.toAnthropicMessages(messages);

    const request: Record<string, unknown> = {
      model: resolvedModel,
      messages: anthropicMessages,
      max_tokens: 4096,
      temperature,
    };

    if (systemPrompt) request.system = systemPrompt;
    if (tools?.length) request.tools = this.toAnthropicTools(tools);

    const response = await client.messages.create(request);

    let contentText = '';
    const toolCalls: ToolCall[] = [];

    for (const block of response.content) {
      if (block.type === 'text') {
        contentText += block.text;
      } else if (block.type === 'tool_use') {
        toolCalls.push({
          id: block.id,
          name: block.name,
          arguments: block.input,
        });
      }
    }

    return {
      role: 'assistant',
      content: contentText,
      toolCalls: toolCalls.length > 0 ? toolCalls : undefined,
    };
  }

  private async *streamAnthropic(
    messages: Message[],
    model: string,
    temperature: number,
  ): AsyncGenerator<StreamEvent> {
    const client = await this.getAnthropicClient() as {
      messages: {
        stream: (req: Record<string, unknown>) => {
          on: (event: string, cb: (e: { type: string; text?: string }) => void) => unknown;
          finalMessage: () => Promise<unknown>;
          [Symbol.asyncIterator](): AsyncIterator<{ type: string; delta?: { text?: string } }>;
        };
      };
    };

    const resolvedModel = this.resolveModel(model, 'anthropic');
    const { systemPrompt, messages: anthropicMessages } = this.toAnthropicMessages(messages);

    const request: Record<string, unknown> = {
      model: resolvedModel,
      messages: anthropicMessages,
      max_tokens: 4096,
      temperature,
    };

    if (systemPrompt) request.system = systemPrompt;

    const stream = client.messages.stream(request);

    for await (const event of stream) {
      if (event.type === 'content_block_delta' && event.delta?.text) {
        yield { type: 'text_delta', data: event.delta.text };
      }
    }

    yield { type: 'done' };
  }

  private toAnthropicMessages(messages: Message[]): {
    systemPrompt: string;
    messages: Record<string, unknown>[];
  } {
    let systemPrompt = '';
    const anthropicMessages: Record<string, unknown>[] = [];

    for (const msg of messages) {
      if (msg.role === 'system') {
        systemPrompt = msg.content;
        continue;
      }

      if (msg.role === 'tool') {
        anthropicMessages.push({
          role: 'user',
          content: [
            {
              type: 'tool_result',
              tool_use_id: msg.toolCallId,
              content: msg.content,
            },
          ],
        });
      } else if (msg.role === 'assistant' && msg.toolCalls?.length) {
        const content: Record<string, unknown>[] = [];
        if (msg.content) content.push({ type: 'text', text: msg.content });
        for (const tc of msg.toolCalls) {
          content.push({
            type: 'tool_use',
            id: tc.id,
            name: tc.name,
            input: tc.arguments,
          });
        }
        anthropicMessages.push({ role: 'assistant', content });
      } else {
        anthropicMessages.push({ role: msg.role, content: msg.content });
      }
    }

    return { systemPrompt, messages: anthropicMessages };
  }

  private toAnthropicTools(tools: Record<string, unknown>[]): Record<string, unknown>[] {
    return tools.map((tool) => ({
      name: tool.name,
      description: tool.description ?? '',
      input_schema: tool.parameters ?? { type: 'object', properties: {} },
    }));
  }

  // --- Google ---

  private async callGoogle(
    messages: Message[],
    model: string,
    temperature: number,
  ): Promise<Message> {
    let genai: { GoogleGenerativeAI: new (key: string) => {
      getGenerativeModel: (opts: { model: string }) => {
        generateContent: (req: { contents: unknown[]; generationConfig: unknown }) => Promise<{
          response: { text: () => string };
        }>;
      };
    }};

    try {
      genai = await dynamicImport('@google/generative-ai') as typeof genai;
    } catch {
      throw new ConfigurationError(
        'Google AI package not installed. Run: npm install @google/generative-ai',
      );
    }

    const resolvedModel = this.resolveModel(model, 'google');
    const client = new genai.GoogleGenerativeAI(this.config.googleApiKey ?? '');
    const genModel = client.getGenerativeModel({ model: resolvedModel });

    // Build simple prompt from messages
    const promptParts = messages.map((msg) => {
      const prefix: Record<string, string> = {
        system: 'System: ',
        user: '',
        assistant: 'Assistant: ',
        tool: 'Tool result: ',
      };
      return `${prefix[msg.role] ?? ''}${msg.content}`;
    });

    const result = await genModel.generateContent({
      contents: [{ role: 'user', parts: [{ text: promptParts.join('\n\n') }] }],
      generationConfig: { temperature },
    });

    return {
      role: 'assistant',
      content: result.response.text() ?? '',
    };
  }

  // --- Ollama ---

  private async callOllama(
    messages: Message[],
    model: string,
    temperature: number,
  ): Promise<Message> {
    let OpenAI: new (opts: { baseURL: string; apiKey: string }) => {
      chat: {
        completions: {
          create: (req: Record<string, unknown>) => Promise<{
            choices: Array<{ message: { content: string | null } }>;
          }>;
        };
      };
    };

    try {
      ({ OpenAI } = await dynamicImport('openai') as any);
    } catch {
      throw new ConfigurationError(
        'OpenAI package needed for Ollama compatibility. Run: npm install openai',
      );
    }

    const host = this.config.ollamaHost ?? 'http://localhost:11434';
    const client = new OpenAI({ baseURL: `${host}/v1`, apiKey: 'ollama' });
    const resolvedModel = this.resolveModel(model, 'ollama');

    const response = await client.chat.completions.create({
      model: resolvedModel,
      messages: this.toOpenAIMessages(messages),
      temperature,
    });

    return {
      role: 'assistant',
      content: response.choices[0].message.content ?? '',
    };
  }
}
